xarray/core/combine.py

```diff
@@ -606,6 +606,21 @@ def vars_as_keys(ds):
     return tuple(sorted(ds))
 
 
+T = TypeVar("T")
+K = TypeVar("K", bound=Hashable)
+
+
+def groupby_defaultdict(
+    iter: list[T], key: Callable[[T], K] = lambda x: x
+) -> Iterator[tuple[K, Iterator[T]]]:
+    """replacement for itertools.groupby"""
+    idx = defaultdict(list)
+    for i, obj in enumerate(iter):
+        idx[key(obj)].append(i)
+    for k, ix in idx.items():
+        yield k, (iter[i] for i in ix)
```

Comment:
@headtr1ck 

Can you make anything out of this:

```
xarray/core/combine.py: note: In function "groupby_defaultdict":
xarray/core/combine.py:614: error: Incompatible default for argument "key" (default has type "Callable[[T], T]", argument has type "Callable[[T], K]")  [assignment]
xarray/core/combine.py:614: error: Incompatible return value type (got "T", expected "K")  [return-value]
```

Reply:
Interesting.

I guess the issue is when you pass a list of non-hashables (like the list of Datasets) and use the default, which ofc would not work.

I would have guessed that the type checker would complain when calling it like that, not in the function definition....

Not sure if that is a mypy bug or a feature. 
Have to think about it more.